server/autolink/lib_jira_test.go

```diff
@@ -80,43 +80,75 @@ var jiraTests = []linkTest{
 		},
 		"Welcome https://mattermost.atlassian.net/browse/MM-12345. should link https://mattermost.atlassian.net/browse/MM-12346 !",
 		"Welcome [MM-12345](https://mattermost.atlassian.net/browse/MM-12345). should link [MM-12346](https://mattermost.atlassian.net/browse/MM-12346) !",
 	}, {
 		"Url replacement multiple times and at beginning",
 		autolink.Autolink{
-			Pattern:  "(https:\\/\\/mattermost.atlassian.net\\/browse\\/)(MM)(-)(?P<jira_id>\\d+)",
-			Template: "[MM-$jira_id](https://mattermost.atlassian.net/browse/MM-$jira_id)",
+			Pattern:  "(https:\\/\\/mattermost.atlassian.net\\/browse\\/)(?P<project_id>\\w+)(-)(?P<jira_id>\\d+)",
+			Template: "[${project_id}-${jira_id}](https://mattermost.atlassian.net/browse/${project_id}-${jira_id})",
 		},
 		"https://mattermost.atlassian.net/browse/MM-12345 https://mattermost.atlassian.net/browse/MM-12345",
 		"[MM-12345](https://mattermost.atlassian.net/browse/MM-12345) [MM-12345](https://mattermost.atlassian.net/browse/MM-12345)",
 	}, {
 		"Url replacement at end",
 		autolink.Autolink{
-			Pattern:  "(https://mattermost.atlassian.net/browse/)(MM)(-)(?P<jira_id>\\d+)",
-			Template: "[MM-$jira_id](https://mattermost.atlassian.net/browse/MM-$jira_id)",
+			Pattern:  "(https://mattermost.atlassian.net/browse/)(?P<project_id>\\w+)(-)(?P<jira_id>\\d+)",
+			Template: "[${project_id}-${jira_id}](https://mattermost.atlassian.net/browse/${project_id}-${jira_id})",
 		},
 		"Welcome https://mattermost.atlassian.net/browse/MM-12345",
 		"Welcome [MM-12345](https://mattermost.atlassian.net/browse/MM-12345)",
+	}, {
+		"Comment url replacement",
+		autolink.Autolink{
+			Pattern:  "(https://mattermost.atlassian.net/browse/)(?P<project_id>\\w+)(-)(?P<jira_id>\\d+)[?](focusedCommentId)(=)(?P<comment_id>\\d+)",
+			Template: "[${project_id}-${jira_id} With Comment #${comment_id}](https://mattermost.atlassian.net/browse/${project_id}-${jira_id}?focusedCommentId=${comment_id})",
+		},
+		"Welcome https://mattermost.atlassian.net/browse/MM-12345?focusedCommentId=10210 should link!",
+		"Welcome [MM-12345 With Comment #10210](https://mattermost.atlassian.net/browse/MM-12345?focusedCommentId=10210) should link!",
+	}, {
+		"Comment url replacement multiple times",
+		autolink.Autolink{
+			Pattern:  "(https://mattermost.atlassian.net/browse/)(?P<project_id>\\w+)(-)(?P<jira_id>\\d+)[?](focusedCommentId)(=)(?P<comment_id>\\d+)",
+			Template: "[${project_id}-${jira_id} With Comment #${comment_id}](https://mattermost.atlassian.net/browse/${project_id}-${jira_id}?focusedCommentId=${comment_id})",
+		},
+		"Welcome https://mattermost.atlassian.net/browse/MM-12345?focusedCommentId=10210. should link https://mattermost.atlassian.net/browse/MM-12346?focusedCommentId=10210 !",
+		"Welcome [MM-12345 With Comment #10210](https://mattermost.atlassian.net/browse/MM-12345?focusedCommentId=10210). should link [MM-12346 With Comment #10210](https://mattermost.atlassian.net/browse/MM-12346?focusedCommentId=10210) !",
```

Comment:
Any reason why we're testing with a space after the exclamation point? We shouldn't assume that users will put a space there.

Reply:
@mickmister if we don't add space it will still work.

Reply:
@sibasankarnayak Right, but the test should reflect the most common use of the software, which I assume would not have a space there.